
# 1. Importing data

## Loading functions
```{r}
rm(list=ls())
gc()
output_path <- "./output/Clara/Tau_Sh_McM/"
output_path <- "./output/Clara/Tau_Sh_McM/Splited&Manual/Outliers_removed/"

```

```{r}
set.seed(123)
image_number <- 1
kegg_organism = "mmu"#dre for Danio rerio, hsa for Homo sapiens, mmu for Mus musculus
species <- 10090 #9606 for Human, 10090 for mouse, 7955 for zebrafish
organism <- "org.Mm.eg.db" #"org.Hs.eg.db"
```


```{r}
source("../code/global_variables.R")
source("../code/01_loading_data.R")
source("../code/03_cleaning_data.R")
source("../code/04_data_analysis.R")
```



```{r}
# Select input data file
prot_data <- readxl::read_xlsx(path = "../rawdata/031425_Clara_lfq_tau_sh_mcm.xlsx",sheet = "proteinGroups")

#Sustitute spaces by "_"
prot_data <- prot_data %>% 
  dplyr::rename_with(~ str_replace_all(., " ", "_"))
```


```{r}
#Data comes from proteinGroup file from MaxQuant
# 1. Filter contaminant and Reverse
prot_data <- prot_data %>% dplyr::filter(is.na(Reverse),is.na(Potential_contaminant) )
# 1. Create ID Column #
# Crear la nueva columna ID extrayendo el primer Uniprot ID
prot_data$ID <- sapply(strsplit(as.character(prot_data$Protein_IDs), ";"), function(x) x[1])


colnames(prot_data)[which(names(prot_data) == "Sequence_length")] <- "protein_length_aa"
colnames(prot_data)[which(names(prot_data) == "Gene_names")] <- "genename"
colnames(prot_data)[which(names(prot_data) == "Fasta_headers")] <- "Description"


prot_data<- prot_data %>% 
  dplyr::select(ID, genename, Description, protein_length_aa, starts_with("LFQ"), starts_with("Intensity"))


```


```{r}
loading_data(prot_data)

```

# 2. Experiment Design
```{r}
# Experiment design ####
# You just need modify script in Experiment heading
comparisons <- c("TAU_vs_SH", "SH_vs_MCM", "TAU_vs_MCM")

label <- c("LFQ_intensity_MCM_2A", "LFQ_intensity_MCM_2B", "LFQ_intensity_MCM_3", "LFQ_intensity_MCM_4",
           "LFQ_intensity_SH_1", "LFQ_intensity_SH_2",  "LFQ_intensity_SH_3", "LFQ_intensity_SH_4",
           "LFQ_intensity_TAU_1", "LFQ_intensity_TAU_2", "LFQ_intensity_TAU_3", "LFQ_intensity_TAU_4")

'label <- c("Intensity_MCM_2A", "Intensity_MCM_2B", "Intensity_MCM_3", "Intensity_MCM_4",
           "Intensity_SH_1", "Intensity_SH_2",  "Intensity_SH_3", "Intensity_SH_4",
           "Intensity_TAU_1", "Intensity_TAU_2", "Intensity_TAU_3", "Intensity_TAU_4")'
label_columns <- label
columns_to_rename <- c("MCM_1", "MCM_2", "MCM_3", "MCM_4",
                       "SH_1", "SH_2",  "SH_3", "SH_4",
                       "TAU_1", "TAU_2", "TAU_3", "TAU_4")

condition <- as.factor(c("MCM","MCM","MCM","MCM",
                         "SH","SH","SH","SH",
                         "TAU","TAU","TAU","TAU"))
  

replicate <- c(1,2,3,4,
               1,2,3,4,
               1,2,3,4)


experiment <- rep("Tau_Mcm_Sh",length(replicate))

Exp_design <- cbind.data.frame(label,condition,replicate,columns_to_rename, experiment)
Exp_design


print(Exp_design)

Exp_design %>% 
  write_xlsx(paste0(output_path,"tables/experiment_desing.xlsx"))

```

## Outliers Removed
```{r eval=FALSE, include=FALSE}

source("../code/global_variables.R")
# Experiment design ####
# You just need modify script in Experiment heading
comparisons <- c("TAU_vs_SH", "SH_vs_MCM", "TAU_vs_MCM")

label <- c("LFQ_intensity_MCM_2A", "LFQ_intensity_MCM_2B", "LFQ_intensity_MCM_3",
           "LFQ_intensity_SH_1", "LFQ_intensity_SH_2",  "LFQ_intensity_SH_3", 
           "LFQ_intensity_TAU_1", "LFQ_intensity_TAU_2", "LFQ_intensity_TAU_4")

label_columns <- label
columns_to_rename <- c("MCM_1", "MCM_2", "MCM_3",
                       "SH_1", "SH_2",  "SH_3", 
                       "TAU_1", "TAU_2", "TAU_4")

condition <- as.factor(c("MCM","MCM","MCM",
                         "SH","SH","SH",
                         "TAU","TAU","TAU"))
  

replicate <- c(1,2,3,
               1,2,3,
               1,2,4)


experiment <- rep("Tau_Mcm_Sh",length(replicate))

Exp_design <- cbind.data.frame(label,condition,replicate,columns_to_rename, experiment)
Exp_design


print(Exp_design)

Exp_design %>% 
  write_xlsx(paste0(output_path,"tables/experiment_desing.xlsx"))

```


# 3. Data Cleaning

When your biology really includes true “off” proteins in one condition (MNAR) but you still want to preserve as much within-group variance as possible, the sweet spot in our diagnostics was a mixed imputation with:

MNAR cutoff (fraction_NA) around 0.6 (i.e. only treat proteins missing in ≥60 % of one condition as left-censored)

Down-shift factor (factor_SD_impute) around 0.05 (so your replacement values sit just below the noise floor, but don’t flatten out the variance)

MAR method = knn, MNAR method = “zero”+manual shift
```{r}
source("../code/03_cleaning_data_mixed_imputation.R")
data_cleaning(prot_data, Exp_design, comparisons, output_path, fraction_NA =0.6, factor_SD_impute =0.05, mnar_var="zero" )
```


# 4. Data Analysis
```{r}
# Load desired imputation file

#data_analysis(data_imp, Exp_design, comparisons, output_path)
#data_analysis(data_imp_man, Exp_design, comparisons, output_path)

data_analysis(mixed_splited_imputation, Exp_design, comparisons, output_path)
```


# 5. Plots
```{r}
#Change to 1 to print Barplots
print_barplots <- 0

source("../code/05_Plots.R")

```

# 6. GO
```{r}
#Use upregulated and downregulated genes independently
independent_UPDOWN <- 1
#Select type of ontology
ont <- "ALL"  # Can be "ALL", "BP", "CC", or "MF"
source("../code/06_GO.R")

```

# 7. Strings
```{r}
independent_UPDOWN <- 1
source("../code/07_Strings.R")
```


# 8. gseGO & KEGG
```{r}
#Use upregulated and downregulated genes independently
independent_UPDOWN <- 1
#Select type of ontology
ont <- "ALL"  # Can be "ALL", "BP", "CC", or "MF"
source("../code/08_gseGO.R")
```


```{r}
source("../code/09_gseKEGG.R")

```

# 10. RBioAPI String and Panther
```{r}
source("../code/10_RBioApi_string.R") 
source("../code/11_RBioApi_panther.R")
```

# 11. Adj

```{r}
source("../code/13_GO_padj.R")

source("../code/15_gseGO_adj.R")
source("../code/13_gseKEGG_padj.R")


```

